Added nullable annotations to Avalonia.Controls.

grokys opened the pull request:
## What does the pull request do?

Adds nullable annotations to Avalonia.Controls and fixes resulting errors in Avalonia.Diagnostics.

This is a big one! I'm not sure I've got everything right here, so if you're responsible for a control/class here and the annotations look wrong to you, please simply push a fix to this branch if you have write access, otherwise comment and I'll fix.

grokys committed:
Add nullable annotations to Avalonia.Controls.

grokys committed:
Fix up nullable issues in Avalonia.Diagnostics.

grokys committed:
Fix nullability error from Avalonia.Web.Blazor.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0018327-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

grokys committed:
Remove #nullable directives.

grokys committed:
Remove usages of Contract.Requires.

grokys committed:
Relax constraint on ITemplate<T>.

maxkatz6 commented:
> if you're responsible for a control/class here and the annotations look wrong to you, please simply push a fix to this branch if you have write access

I am pretty sure we will come back to this after merging this PR anyway.

maxkatz6 commented on src/Avalonia.Controls/Platform/ITopLevelImplWithTextInputMethod.cs:
```diff
@@ -6,6 +6,6 @@ namespace Avalonia.Controls.Platform
     public interface ITopLevelImplWithTextInputMethod : ITopLevelImpl
     {
-        public ITextInputMethodImpl TextInputMethod { get; }
+        public ITextInputMethodImpl? TextInputMethod { get; }
```
Looks a bit wrong. Why is it nullable, when ITopLevelImplWithTextInputMethod is special interface specifically to provide this property?

grokys replied:
Because `ValidatingTopLevel` uses an `as` cast here:

https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/ValidatingToplevel.cs#L131

@kekekeks is this correct, or should that be a throwing cast?

kekekeks replied:
It's nullable because it might be not available for this particular toplevel. E. g. if we are running on Linux without any configured IME

maxkatz6 commented on src/Avalonia.Controls/Platform/ITopLevelNativeMenuExporter.cs:
```diff
@@ -24,6 +22,6 @@ public interface INativeMenuExporterProvider
     public interface ITopLevelImplWithNativeMenuExporter : ITopLevelImpl
     {
-        ITopLevelNativeMenuExporter NativeMenuExporter { get; }
+        ITopLevelNativeMenuExporter? NativeMenuExporter { get; }
```
Same

grokys replied:
Because of the `?` here:

https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/TrayIcon.cs#L187

@danwalmsley is this correct?

kekekeks replied:
It's nullable because it might be not available for this particular toplevel. E. g. if we are running on Linux with DE that does not support the global menu. Or if global menu was disabled in platform settings (Linux, macOS)

maxkatz6 commented on src/Avalonia.Controls/Presenters/ScrollContentPresenter.cs:
```diff
@@ -481,9 +479,10 @@ protected override void OnPropertyChanged<T>(AvaloniaPropertyChangedEventArgs<T>
+        private void BringIntoViewRequested(object? sender, RequestBringIntoViewEventArgs e)
         {
-            e.Handled = BringDescendantIntoView(e.TargetObject, e.TargetRect);
+            if (e.TargetObject is not null)
```
A bit weird that TargetObject can be null when we want to bring it into the view

grokys replied:
Yeah this is mainly due to the fact that a `RequestBringIntoViewEventArgs` event can be sent by anyone, and they're not guaranteed to have set the `TargetObject` property. In that case it's either throw or ignore, and throwing deep in a bubbling/tunneling event handler isn't a great idea in general.

maxkatz6 commented on src/Avalonia.Controls/Primitives/AccessText.cs:
```diff
@@ -111,8 +111,13 @@ protected override void OnDetachedFromVisualTree(VisualTreeAttachmentEventArgs e
         /// <param name="text">The text.</param>
         /// <returns>The text with the first underscore stripped.</returns>
-        private string StripAccessKey(string text)
+        private string? StripAccessKey(string? text)
```
[return: NotNullWhenNotNull] here?

maxkatz6 replied:
Though probably not worth it

grokys replied:
Might as well I guess, though yeah it's rather academic.

maxkatz6 commented on src/Avalonia.Controls/Templates/FuncTreeDataTemplate.cs:
```diff
@@ -58,7 +58,7 @@ public class FuncTreeDataTemplate : FuncDataTemplate, ITreeDataTemplate
         /// <returns>The child items, or null if no child items.</returns>
         public InstancedBinding ItemsSelector(object item)
         {
-            return InstancedBinding.OneTime(this?._itemsSelector(item));
```
I wonder what's the reason of such weird piece of code was.

grokys replied:
I guess I must have just got off working with js ;) But to be honest, no idea - either a brainfart or copy/replace gone wrong?

maxkatz6 reviewed (approved):
Two sits review done. If I haven't missed anything (I probably have), it looks good.
Only some minor comments.

And merge conflicts are there.

grokys committed:
Merge branch 'master' into fixes/avalonia-controls-nullability

grokys committed:
Handle nullability in new TextBox/TextPresenter.

Made the decision to return a null `TextLayout` when `TextPresenter.Foreground` is null for consistency with our other usages of null brushes, though not 100% sure this is correct.

grokys committed:
Added remaining nullable annotations after merge.

grokys commented on src/Avalonia.Controls/Repeater/ItemsRepeater.cs:
```diff
@@ -705,12 +705,17 @@ private void OnItemTemplateChanged(IDataTemplate oldValue, IDataTemplate newValu
+            ItemTemplateShim = newValue switch
+            {
+                IElementFactory factory => factory,
+                null => null,
```
`ItemTemplateWrapper` would have thrown an NRE in the previous code if `newValue` is null, and `ItemTemplateShim` is nullable anyway, so I figured this is a slightly better way to handle this gracefully.

grokys committed:
Clarify the return value of `StripAccessKey`.

grokys commented:
@maxkatz6 you're a hero for wading through this! 🥇 

Responded to your comments and fixed merge conflicts. We just need a quick bit of input from @kekekeks and @danwalmsley  on a couple of those.

@Gillibald please let me know if the null handling of `TextPresenter.Foreground` makes sense.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0018413-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]
